receiver/splunkhecreceiver/receiver.go

```diff
@@ -0,0 +1,305 @@
+		var err error
+		bodyReader, err = gzip.NewReader(bodyReader)
+		if err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errGzipReaderRespBody, err)
+			return
+		}
+	}
+
+	body, err := ioutil.ReadAll(bodyReader)
+	if err != nil {
+		r.failRequest(ctx, resp, http.StatusBadRequest, errReadBodyRespBody, err)
+		return
+	}
+	if len(body) == 0 {
+		resp.Write(okRespBody)
+		return
+	}
+
+	messagesReceived := 0
+	for _, payload := range bytes.Split(body, []byte("\r\n\r\n")) {
+		if len(payload) == 0 {
+			continue
+		}
+
+		var msg splunkhecexporter.SplunkMetric
+		if err = json.Unmarshal(payload, &msg); err != nil {
+			r.failRequest(ctx, resp, http.StatusBadRequest, errUnmarshalBodyRespBody, err)
+			return
+		}
+		messagesReceived += len(msg.GetValues())
+
+		md, _ := SplunkHecToMetricsData(r.logger, &msg)
+
+		if r.config.AccessTokenPassthrough {
+			if accessToken := req.Header.Get("Splunk"); accessToken != "" {
+				if md.Resource == nil {
+					md.Resource = &resourcepb.Resource{}
+				}
+				if md.Resource.Labels == nil {
+					md.Resource.Labels = make(map[string]string, 1)
+				}
+				md.Resource.Labels["com.splunk.hec.access_token"] = accessToken
+			}
+		}
+		err = r.nextConsumer.ConsumeMetricsData(ctx, *md)
+		if err != nil {
+			break
+		}
+	}
+
+	obsreport.EndMetricsReceiveOp(
+		ctx,
+		typeStr,
+		messagesReceived,
+		messagesReceived,
+		err)
+	if err != nil {
+		r.failRequest(ctx, resp, http.StatusInternalServerError, errNextConsumerRespBody, err)
+	} else {
+		resp.WriteHeader(http.StatusAccepted)
+		resp.Write(okRespBody)
+	}
+}
+
+func (r *splunkReceiver) failRequest(
+	ctx context.Context,
+	resp http.ResponseWriter,
+	httpStatusCode int,
+	jsonResponse []byte,
+	err error,
+) {
+	resp.WriteHeader(httpStatusCode)
+	if len(jsonResponse) > 0 {
+		// The response needs to be written as a JSON string.
+		_, writeErr := resp.Write(jsonResponse)
+		if writeErr != nil {
+			r.logger.Warn(
+				"Error writing HTTP response message",
+				zap.Error(writeErr),
+				zap.String("receiver", r.config.Name()))
+		}
+	}
+
+	// Use the same pattern as strings.Builder String().
+	msg := *(*string)(unsafe.Pointer(&jsonResponse))
+
+	reqSpan := trace.FromContext(ctx)
+	reqSpan.AddAttributes(
+		trace.Int64Attribute(conventions.AttributeHTTPStatusCode, int64(httpStatusCode)),
+		trace.StringAttribute(conventions.AttributeHTTPStatusText, msg))
+	traceStatus := trace.Status{
+		Code: trace.StatusCodeInvalidArgument,
+	}
+	if httpStatusCode == http.StatusInternalServerError {
+		traceStatus.Code = trace.StatusCodeInternal
+	}
+	if err != nil {
+		traceStatus.Message = err.Error()
+	}
+	reqSpan.SetStatus(traceStatus)
+	reqSpan.End()
+
+	r.logger.Debug(
+		"Splunk HEC receiver request failed",
+		zap.Int("http_status_code", httpStatusCode),
+		zap.String("msg", msg),
+		zap.Error(err), // It handles nil error
+		zap.String("receiver", r.config.Name()))
+}
+
+func initJSONResponse(s string) []byte {
+	respBody, err := json.Marshal(s)
```

Comment:
Why are we json marshaling a string?

Reply:
I copied this from signalfxreceiver. It's used at init time to set up the JSON responses. I think the panic call in here is to ensure that the JSON string passed in is valid JSON, and all JSON values are constants.